crates/rustc_codegen_spirv/src/linker/param_weakening.rs

```diff
@@ -0,0 +1,114 @@
+            // callee parameters we already determined to be unused, because
+            // those parameters (and matching arguments) will get removed later.
+            let (operands, ignore_operands) = if inst.class.opcode == Op::FunctionCall {
+                (
+                    &inst.operands[1..],
+                    unused_params_per_func_id.get(&inst.operands[0].unwrap_id_ref()),
+                )
+            } else {
+                (&inst.operands[..], None)
+            };
+
+            for (i, operand) in operands.iter().enumerate() {
+                if let Some(ignore_operands) = ignore_operands {
+                    if ignore_operands.contains(i) {
+                        continue;
+                    }
+                }
+
+                if let Operand::IdRef(id) = operand {
+                    if let Some(&param_idx) = params_id_to_idx.get(id) {
+                        unused_params.remove(param_idx);
+                    }
+                }
+            }
+        }
+
+        if !unused_params.is_empty() {
+            unused_params_per_func_id.insert(func.def_id().unwrap(), unused_params);
+        }
+    }
+
+    // Remove unused parameters and call arguments for unused parameters.
+    let mut builder = Builder::new_from_module(module);
+    for func_idx in 0..builder.module_ref().functions.len() {
+        let func = &mut builder.module_mut().functions[func_idx];
+        let unused_params = unused_params_per_func_id.get(&func.def_id().unwrap());
+        if let Some(unused_params) = unused_params {
+            func.parameters = mem::take(&mut func.parameters)
+                .into_iter()
+                .enumerate()
+                .filter(|&(i, _)| !unused_params.contains(i))
+                .map(|(_, p)| p)
+                .collect();
+        }
+
+        for inst in func.all_inst_iter_mut() {
+            if inst.class.opcode == Op::FunctionCall {
+                if let Some(unused_callee_params) =
+                    unused_params_per_func_id.get(&inst.operands[0].unwrap_id_ref())
+                {
+                    inst.operands = mem::take(&mut inst.operands)
+                        .into_iter()
+                        .enumerate()
+                        .filter(|&(i, _)| i == 0 || !unused_callee_params.contains(i - 1))
+                        .map(|(_, o)| o)
+                        .collect();
+                }
+            }
+        }
+
+        // Regenerate the function type from remaining parameters, if necessary.
+        if unused_params.is_some() {
+            let return_type = func.def.as_mut().unwrap().result_type.unwrap();
+            let new_param_types: Vec<_> = func
+                .parameters
+                .iter()
+                .map(|inst| inst.result_type.unwrap())
+                .collect();
+            let new_func_type = builder.type_function(return_type, new_param_types);
```

Comment:
Hrmpf. `O(n^2)` linear search here for dedup isn't great, but I guess it's fine, usually not many functions.

Reply:
"usually not many" - famous last words ;)